File: tests/test_catalog_sync_health.py
"""Offline monitor regressions: genuine outages must never become healthy reports."""
import contextlib
from datetime import datetime, timezone
import io
from pathlib import Path
import sys
import unittest
from unittest.mock import patch

import requests

sys.path.insert(0, str(Path(__file__).resolve().parents[1] / "scripts"))
import check_catalog_sync as monitor


def fresh(**changes):
    stamp = datetime.now(timezone.utc).isoformat()
    return {"available": True, "report_stale": False, "sync_stale": False,
            "reported_at": stamp, "last_sync_at": stamp,
            "local_papers": 100, "synced_papers": 80, "citation_pending": 20,
            "local_originals": 30, "local_summaries": 10,
            "pending_originals": 3, "pending_summaries": 2,
            "sync_state": "idle", "registry_version": "fixture-v1", **changes}


class CatalogSyncHealthTests(unittest.TestCase):
    def run_monitor(self, status=None, error=None):
        output = io.StringIO()
        with patch.dict(monitor.os.environ, {"SUPABASE_URL": "https://fixture.example/",
                                            "SUPABASE_SERVICE_KEY": "sb_secret_fixture"}), \
                patch.object(monitor, "get_json", return_value=status, side_effect=error) as get, \
                contextlib.redirect_stdout(output):
            monitor.main()
        return output.getvalue(), get

    def test_fresh_report_uses_only_lightweight_rpc(self):
        output, get = self.run_monitor(fresh())
        get.assert_called_once_with("https://fixture.example/rest/v1/rpc/catalog_sync_health",
                                    headers={"apikey": "sb_secret_fixture"}, params={})
        self.assertIn("Current revisions synchronized: 80", output)
        self.assertNotIn("::warning::", output)

    def test_capacity_blocking_is_warning_not_false_completion(self):
        output, _ = self.run_monitor(fresh(sync_state="capacity_blocked"))
        self.assertIn("::warning::", output)
        self.assertIn("uploads remain queued", output)
        self.assertIn("Citation synchronization pending: 20", output)

    def test_active_cycle_with_recent_success_can_be_healthy(self):
        self.assertEqual(monitor.validate_health(fresh(sync_state="syncing"))["sync_state"], "syncing")

    def test_missing_stale_and_error_reports_fail(self):
        for status in [
            {"available": False, "report_stale": True, "sync_stale": True},
            fresh(report_stale=True), fresh(sync_stale=True, last_sync_at=None),
            fresh(sync_stale=True, last_sync_at="2020-01-01T00:00:00+00:00"),
            fresh(sync_state="offline"), fresh(sync_state="error"),
        ]:
            with self.subTest(status=status), self.assertRaises(SystemExit) as caught:
                self.run_monitor(status)
            self.assertIn("health failed", str(caught.exception))

    def test_unknown_states_and_malformed_counters_fail_closed(self):
        for status in [None, [], {}, fresh(available=1), fresh(report_stale="false"),
                       fresh(sync_stale=None), fresh(sync_state="healthy"), fresh(sync_state=[]),
                       fresh(local_papers=True), fresh(local_papers="100"), fresh(local_papers=-1),
                       fresh(local_papers=1_000_000_001), fresh(citation_pending=19),
                       fresh(pending_originals=31), fresh(pending_summaries=101),
                       fresh(local_originals=101), fresh(local_summaries=101),
                       fresh(registry_version=None), fresh(registry_version="")]:
            with self.subTest(status=status), self.assertRaises(ValueError):
                monitor.validate_health(status)

    def test_invalid_timestamp_does_not_pass_even_with_fresh_flags(self):
        for field in ("reported_at", "last_sync_at"):
            for stamp in (None, 123, "invalid", "2026-01-01T00:00:00"):
                with self.subTest(field=field, stamp=stamp), self.assertRaisesRegex(ValueError, "timestamp"):
                    monitor.validate_health(fresh(**{field: stamp}))

    def test_request_failure_remains_failed_and_does_not_expose_exception(self):
        for error in (requests.Timeout("secret-token response-body"),
                      requests.HTTPError("https://secret-token@fixture.example/private"),
                      requests.RequestException("secret-token response-body"),
                      ValueError("secret-token invalid JSON response-body")):
            with self.subTest(error=type(error).__name__), self.assertRaises(SystemExit) as caught:
                self.run_monitor(error=error)
            self.assertIn("cloud API request failed", str(caught.exception))
            self.assertNotIn("secret-token", str(caught.exception))
            self.assertTrue(caught.exception.__suppress_context__)

    def test_missing_configuration_fails_before_any_request(self):
        with patch.dict(monitor.os.environ, {}, clear=True), patch.object(monitor, "get_json") as get:
            with self.assertRaisesRegex(SystemExit, "required"):
                monitor.main()
            get.assert_not_called()


if __name__ == "__main__":
    unittest.main()

File: tests/test_catalog_sync_health.mjs
// Isolated PostgreSQL health contract; no production calls or credentials.
import assert from 'node:assert/strict';
import {createHash} from 'node:crypto';
import {readFile,readdir} from 'node:fs/promises';
import {PGlite} from '../frontend/node_modules/@electric-sql/pglite/dist/index.js';
const db=new PGlite();
const worker='00000000-0000-0000-0000-000000000009',other='00000000-0000-0000-0000-000000000010';
const owner='00000000-0000-0000-0000-000000000001',reader='00000000-0000-0000-0000-000000000002';
const token='health-fixture-token-'.repeat(3);
const scalar=async(sql,params=[])=> (await db.query(sql,params)).rows[0].r;
const health=()=>scalar('SELECT public.catalog_sync_health() r');
const status={local_papers:100,synced_papers:80,citation_pending:20,local_originals:30,
 local_summaries:10,pending_originals:3,pending_summaries:2,sync_state:'capacity_blocked',
 last_sync_at:new Date().toISOString(),registry_version:'fixture-v1'};
const report=(value,id=worker)=>db.query('SELECT public.report_institution_catalog($1,$2,$3)',[id,token,JSON.stringify(value)]);
try {
 await db.exec(`CREATE ROLE anon;CREATE ROLE authenticated;CREATE ROLE service_role BYPASSRLS;CREATE SCHEMA auth;
 CREATE TABLE auth.users(id uuid PRIMARY KEY,email text,email_confirmed_at timestamptz,raw_user_meta_data jsonb);
 CREATE FUNCTION auth.uid() RETURNS uuid LANGUAGE sql AS $$SELECT nullif(current_setting('request.jwt.claim.sub',true),'')::uuid$$;
 GRANT USAGE ON SCHEMA auth,public TO authenticated,anon;`);
 for(const file of (await readdir(new URL('../supabase/migrations/',import.meta.url))).filter(f=>f.endsWith('.sql')).sort())
  await db.exec((await readFile(new URL('../supabase/migrations/'+file,import.meta.url),'utf8')).replace(/^\uFEFF/,''));
 await db.query('INSERT INTO app_private.institution_workers(id,name,token_hash) VALUES($1,$2,$3),($4,$5,$3)',
  [worker,'Fixture',createHash('sha256').update(token).digest('hex'),other,'Second']);
 await db.query("INSERT INTO auth.users VALUES($1,'crazyslime@gmail.com',now(),'{}'),($2,'reader@example.test',now(),'{}')",[owner,reader]);
 await db.exec('SET ROLE service_role');
 assert.deepEqual(await health(),{available:false,report_stale:true,sync_stale:true});
 await db.exec('RESET ROLE');await report(status);
 await db.exec('SET ROLE service_role');
 let current=await health();
 assert.equal(current.available,true);assert.equal(current.report_stale,false);assert.equal(current.sync_stale,false);
 assert.equal(current.sync_state,'capacity_blocked');assert.equal(current.citation_pending,20);
 assert.equal('worker_id' in current,false);assert.equal('token_hash' in current,false);
 await db.exec("RESET ROLE;UPDATE app_private.institution_catalog_reports SET reported_at=now()-interval '3 hours'");
 current=await health();assert.equal(current.report_stale,true);assert.equal(current.sync_stale,false);
 await db.exec("UPDATE app_private.institution_catalog_reports SET reported_at=now()+interval '10 minutes'");
 assert.equal((await health()).report_stale,true,'Future reports cannot extend freshness');
 for(const last_sync_at of [null,'2020-01-01T00:00:00+00:00']) {
  await report({...status,last_sync_at});current=await health();
  assert.equal(current.report_stale,false);assert.equal(current.sync_stale,true,'Recent report must not hide absent/old cycle');
 }
 // Malformed legacy/manual rows do not turn fresh. Token publication already
 // rejects malformed/future timestamps, tested by the publication suite.
 for(const value of ['invalid','infinity','2099-01-01T00:00:00+00:00',123,{}]) {
  await db.query("UPDATE app_private.institution_catalog_reports SET status=jsonb_set(status,'{last_sync_at}',$1)",[JSON.stringify(value)]);
  assert.equal((await health()).sync_stale,true);
 }
 await report(status);
 await db.exec("UPDATE app_private.institution_catalog_reports SET reported_at=now()-interval '1 minute'");
 await report({...status,local_papers:200,citation_pending:120},other);
 assert.equal((await health()).local_papers,200,'Newest enabled worker provides the report');
 await db.query('UPDATE app_private.institution_workers SET enabled=false WHERE id=$1',[other]);
 assert.equal((await health()).local_papers,100,'Disabled workers cannot mask a stale or missing active report');
 await db.query('UPDATE app_private.institution_workers SET enabled=false WHERE id=$1',[worker]);
 assert.equal((await health()).available,false);
 await db.query('UPDATE app_private.institution_workers SET enabled=true WHERE id=$1',[worker]);
 for(const state of ['offline','error','syncing']) {
  await report({...status,sync_state:state});assert.equal((await health()).sync_state,state,'RPC preserves failures for the monitor');
 }
 await db.exec('SET ROLE anon');await assert.rejects(health(),{code:'42501'});
 await assert.rejects(db.query('SELECT * FROM app_private.institution_catalog_reports'),{code:'42501'});
 await db.exec('RESET ROLE;SET ROLE authenticated');
 for(const id of [reader,owner]) {
  await db.query("SELECT set_config('request.jwt.claim.sub',$1,false)",[id]);
  await assert.rejects(health(),{code:'42501'},'Direct health RPC remains service-only, including admin accounts');
 }
 assert.equal((await scalar('SELECT public.admin_catalog_status() r')).local_catalog.local_papers,100,'Existing admin wrapper remains available');
 await db.query("SELECT set_config('request.jwt.claim.sub',$1,false)",[reader]);
 await assert.rejects(db.query('SELECT public.admin_catalog_status()'),{code:'42501'});
 await db.exec(`RESET ROLE;
 CREATE OR REPLACE FUNCTION public.catalog_backfill_status_v26() RETURNS jsonb
 LANGUAGE plpgsql SECURITY DEFINER SET search_path='' AS $$ BEGIN RAISE EXCEPTION 'aggregate sentinel'; END $$;
 SET ROLE service_role;`);
 assert.equal((await health()).local_papers,100,'Heartbeat does not execute catalog aggregates');
 assert.equal((await scalar('SELECT public.catalog_backfill_status() r')).local_catalog.local_papers,100,
  'Administrative status also avoids the obsolete wide aggregate after metrics migration');
 console.log('Catalog sync health: lightweight report access, report/cycle freshness, enabled-worker selection and service/admin boundaries passed.');
} finally {await db.close();}
